components/site/patterns/DeleteTable.js

```diff
@@ -5,59 +5,68 @@ import { Button } from '@twilio-paste/core/button';
 import { DeleteIcon } from '@twilio-paste/icons/cjs/DeleteIcon';
 import PropTypes from 'prop-types';
 
-export const DeleteTable = ({ data, handleDelete }) => {
-  if (data.length === 0) {
+export const DeleteTable = ({ columns, rows, emptyState, handleDelete }) => {
+  if (rows.length === 0) {
     return (
       <Text as="span" fontFamily="fontFamilyText">
-        Nothing to delete!
+        {emptyState}
       </Text>
     );
   }
 
-  const cols = Object.keys(data[0]);
+  rows.forEach((row) => {
+    if (row.length > columns.length)
+      throw new Error(
+        'If you are seeing this error, it is because you added more cells than there are columns. Go back to delete.js and add a new column heading. DO NOT try to fix this error in DeleteTable.js!',
+      );
+  });
 
   return (
     <Table>
       <THead>
         <Tr>
-          {cols.map((key) => {
-            if (key === 'id') return <></>;
-            return <Th key={cols.indexOf(key)}>{key}</Th>;
+          {columns.map((column) => {
+            return <Th key={columns.indexOf(column)}>{column}</Th>;
           })}
           <Th>Actions</Th>
         </Tr>
       </THead>
       <TBody>
-        {data.map((service) => {
-          const row = [];
-          for (let i = 1; i < cols.length; i++) {
-            row.push(<Td key={service.id + i}>{service[cols[i]]}</Td>);
+        {rows.map((row, index) => {
+          if (row.length < columns.length) {
+            const diff = columns.length - row.length;
+            for (let i = 0; i < diff; i++) {
+              row.push(<Text as="span" fontFamily="fontFamilyText" />);
+            }
           }
-          row.push(
-            <Td key={service.id}>
-              <Button
-                id={service.id}
-                size="icon_small"
-                variant="destructive_secondary"
-                onClick={() => handleDelete(service)}
-              >
-                <DeleteIcon title="Delete" />
-              </Button>
-            </Td>,
+          return (
+            <Tr key={index}>
+              {row.map((cell, idx) => (
+                <Td key={row[0].props.children + idx}>{cell}</Td>
+              ))}
+              <Td key={index}>
+                <Button id={index} size="icon_small" variant="destructive_secondary" onClick={() => handleDelete(row)}>
+                  <DeleteIcon title="Delete" />
+                </Button>
+              </Td>
+            </Tr>
```

Comment:
So when you are mapping over the rows here, one approach might be to loop/map over the columns to render the cells so that you always have the correct number of columns in the table.

Whilst looping through the columns, you can use the index to pull the corresponding cell from the row. Simple example

```jsx
{rows.map((row, index) => (
  <Tr key={index}>
    {columns.map((column, index) => {
      return (
        <Td key={index}>{row[index]}</Td>
      )
    })}
  </Tr>
))}
```

(typed on a browser, syntax likely wrong)

Reply:
wow works great!! implemented this to include the delete button. tested that nothing breaks when adding a new column and still receive error when adding a row without a column👍